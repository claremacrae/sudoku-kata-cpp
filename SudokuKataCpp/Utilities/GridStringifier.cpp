#include "GridStringifier.h"

namespace SudokuKata::Utilities
{

    std::string GridStringifier::ConvertToString(const std::vector<int>& board)
    {
        const std::string divider = "+---+---+---+\n";
        auto result = divider;
        result += ToRowNumbered(board, 0);
        result += ToRowNumbered(board, 1);
        result += ToRowNumbered(board, 2);
        result += divider;
        result += ToRowNumbered(board, 3);
        result += ToRowNumbered(board, 4);
        result += ToRowNumbered(board, 5);
        result += divider;
        result += ToRowNumbered(board, 6);
        result += ToRowNumbered(board, 7);
        result += ToRowNumbered(board, 8);
        result += divider;
        return result;
    }

    std::string GridStringifier::ToRowNumbered(const std::vector<int>& board, int row)
    {
        auto o = row * 9;
        return ToRow(board, o + 0, o + 1, o + 2, o + 3, o + 4, o + 5, o + 6, o + 7, o + 8);
    }

    std::string GridStringifier::ToRow(const std::vector<int>& board,
                                       int index0,
                                       int index1,
                                       int index2,
                                       int index3,
                                       int index4,
                                       int index5,
                                       int index6,
                                       int index7,
                                       int index8)
    {
        std::string row = "|";
        row += ToTriplet(board, index0, index1, index2);
        row += "|";
        row += ToTriplet(board, index3, index4, index5);
        row += "|";
        row += ToTriplet(board, index6, index7, index8);
        row += "|";
        row += "\n";
        return row;
    }

    std::string
    GridStringifier::ToTriplet(const std::vector<int>& board, int index0, int index1, int index2)
    {
        auto value1 = board[index0];
        auto value2 = board[index1];
        auto value3 = board[index2];
        return ToCell(value1) + ToCell(value2) + ToCell(value3);
    }

    std::string GridStringifier::ToCell(int value)
    {
        return value == 0 ? "." : std::to_string(value);
    }

    std::string GridStringifier::ConvertToCode(const std::vector<int>& board)
    {
        std::string result;
        for (auto digit : board)
        {
            result += std::to_string(digit);
        }
        return result;
    }
}
