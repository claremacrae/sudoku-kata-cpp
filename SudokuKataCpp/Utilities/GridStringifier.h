#pragma once

#include <string>
#include <vector>

namespace SudokuKata::Utilities
{
    class GridStringifier
    {
    public:
        static std::string ConvertToString(const std::vector<int>& board);

    private:
        static std::string ToRowNumbered(const std::vector<int>& board, int row);

        static std::string ToRow(const std::vector<int>& board,
                                 int index0,
                                 int index1,
                                 int index2,
                                 int index3,
                                 int index4,
                                 int index5,
                                 int index6,
                                 int index7,
                                 int index8);

        static std::string ToTriplet(const std::vector<int>& board, int index0, int index1, int index2);

        static std::string ToCell(int value);

    public:
        static std::string ConvertToCode(const std::vector<int>& board);
    };
}
