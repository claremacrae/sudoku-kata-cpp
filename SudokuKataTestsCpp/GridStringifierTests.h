#pragma once

#include <vector>

namespace SudokuKataTests
{
    class GridStringifierTests
    {
    public:
        static void VerifyBoardAsString(std::vector<int>& input);

        static void VerifyBoardAsCode(std::vector<int>& input);
    };
}
