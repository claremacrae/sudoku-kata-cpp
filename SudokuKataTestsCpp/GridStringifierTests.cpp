#include "GridStringifierTests.h"

#include "catch2/catch.hpp"
#include "ApprovalTests.hpp"

#include "Utilities/GridStringifier.h"

using namespace SudokuKataTests;

using namespace ApprovalTests;

TEST_CASE("TestPopulatedBoard")
{
    // Assert
    std::vector<int> input = {4, 2, 9, 5, 3, 8, 6, 1, 7, 6, 3, 7, 9, 1, 4, 8, 2, 5, 8, 5, 1,
                              7, 6, 2, 4, 3, 9, 9, 7, 5, 6, 8, 3, 1, 4, 2, 1, 6, 2, 4, 5, 7,
                              9, 8, 3, 3, 4, 8, 1, 2, 9, 5, 7, 6, 2, 1, 6, 8, 7, 5, 3, 9, 4,
                              5, 9, 3, 2, 4, 1, 7, 6, 8, 7, 8, 4, 3, 9, 6, 2, 5, 1};
    GridStringifierTests::VerifyBoardAsString(input);
}

TEST_CASE("TestPopulatedBoardAsCode")
{
    // Assert
    std::vector<int> input = {4, 2, 9, 5, 3, 8, 6, 1, 7, 6, 3, 7, 9, 1, 4, 8, 2, 5, 8, 5, 1,
                              7, 6, 2, 4, 3, 9, 9, 7, 5, 6, 8, 3, 1, 4, 2, 1, 6, 2, 4, 5, 7,
                              9, 8, 3, 3, 4, 8, 1, 2, 9, 5, 7, 6, 2, 1, 6, 8, 7, 5, 3, 9, 4,
                              5, 9, 3, 2, 4, 1, 7, 6, 8, 7, 8, 4, 3, 9, 6, 2, 5, 1};
    GridStringifierTests::VerifyBoardAsCode(input);
}

TEST_CASE("TestPartialBoard")
{
    // Assert
    std::vector<int> input = {0, 0, 0, 5, 0, 8, 0, 0, 7, 6, 0, 7, 0, 1, 0, 0, 2, 0, 0, 5, 0,
                              0, 0, 0, 0, 3, 9, 0, 7, 0, 6, 8, 0, 1, 0, 0, 0, 6, 0, 0, 5, 7,
                              0, 0, 3, 0, 0, 8, 1, 0, 0, 0, 7, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0,
                              5, 9, 0, 0, 4, 1, 7, 0, 0, 0, 0, 0, 3, 0, 0, 0, 5, 1};
    GridStringifierTests::VerifyBoardAsString(input);
}

TEST_CASE("TestPartialBoardAsCode")
{
    // Assert
    std::vector<int> input = {0, 0, 0, 5, 0, 8, 0, 0, 7, 6, 0, 7, 0, 1, 0, 0, 2, 0, 0, 5, 0,
                              0, 0, 0, 0, 3, 9, 0, 7, 0, 6, 8, 0, 1, 0, 0, 0, 6, 0, 0, 5, 7,
                              0, 0, 3, 0, 0, 8, 1, 0, 0, 0, 7, 0, 0, 0, 6, 0, 0, 0, 0, 0, 0,
                              5, 9, 0, 0, 4, 1, 7, 0, 0, 0, 0, 0, 3, 0, 0, 0, 5, 1};
    GridStringifierTests::VerifyBoardAsCode(input);
}

namespace SudokuKataTests
{
    void GridStringifierTests::VerifyBoardAsString(std::vector<int>& input)
    {
        Approvals::verify(SudokuKata::Utilities::GridStringifier::ConvertToString(input));
    }

    void GridStringifierTests::VerifyBoardAsCode(std::vector<int>& input)
    {
        Approvals::verify(SudokuKata::Utilities::GridStringifier::ConvertToCode(input));
    }
}
