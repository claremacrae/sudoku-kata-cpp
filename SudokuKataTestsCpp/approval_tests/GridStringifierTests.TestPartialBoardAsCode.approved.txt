000508007607010020050000039070680100060057003008100070006000000590041700000300051
